pkg/smokescreen/metrics.go

```diff
@@ -0,0 +1,99 @@
+package smokescreen
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+)
+
+var metrics = map[string][]string{
+	"acl.allow":                        {},
+	"acl.decide_error":                 {},
+	"acl.deny":                         {},
+	"acl.report":                       {},
+	"acl.role_not_determined":          {},
+	"acl.unknown_error":                {},
+	"cn.atpt.connect.time":             {},
+	"cn.atpt.fail.total":               {},
+	"cn.atpt.success.total":            {},
+	"cn.atpt.total":                    {},
+	"resolver.allow.default":           {},
+	"resolver.allow.user_configured":   {},
+	"resolver.attempts_total":          {},
+	"resolver.deny.not_global_unicast": {},
+	"resolver.deny.private_range":      {},
+	"resolver.deny.user_configured":    {},
+	"resolver.errors_total":            {},
+}
+
+// MetricsClient is a thin wrapper around statsd.ClientInterface. It is used to allow
+// adding arbitrary tags to Smokescreen metrics.
+type MetricsClient struct {
+	additionalTags map[string][]string
+	StatsdClient   statsd.ClientInterface
+}
+
+// NewMetricsClient creates a new MetricsClient with the provided statsd address and
+// namespace.
+func NewMetricsClient(addr, namespace string) (*MetricsClient, error) {
+	c, err := statsd.New(addr)
+	if err != nil {
+		return nil, err
+	}
+	c.Namespace = namespace
+
+	return &MetricsClient{
+		additionalTags: make(map[string][]string),
+		StatsdClient:   c,
+	}, nil
+}
+
+// NewNoOpMetricsClient returns a MetricsClient with a no-op statsd client. This can
+// be used when there's no statsd service available to smokescreen.
+func NewNoOpMetricsClient() *MetricsClient {
+	return &MetricsClient{
+		additionalTags: make(map[string][]string),
+		StatsdClient:   &statsd.NoOpClient{},
+	}
+}
+
+// AddMetricTags associates the provided tags slice with a given metric. The metric must be present
+// in the metrics slice.
+func (mc *MetricsClient) AddMetricTags(metric string, mTags []string) error {
+	if tags, ok := metrics[metric]; ok {
+		metrics[metric] = append(tags, mTags...)
```

Comment:
Are there any concurrency concerns around modifying a global?  Maybe this should just be a field inside the MetricsClient instead, along with possibly a lock?

Reply:
Yeah totally agree about moving this inside the `MetricsClient` struct. The intent is that persistent metrics should only be added when the configuration is being initialized, and that `MetricsClient` isn't intended to be thread safe.

Given how concurrent we run smokescreen, I'd be concerned about lock contention if we dropped a mutex in there and would even prefer to avoid an RWLock if possible.

I added comments clarifying the thread safety and intent around MetricsClient.

Reply:
I also added a guard that gets set when smokescreen starts to prevent adding new metrics tags.